### DISCUSSION TITLE
password flags for `*.zip` not took by `7z` from `Expand-7zipArchive`

### DISCUSSION THREAD

**Discussion by goyalyashpal** [ScoopInstaller/Scoop#5866] [Help and Questions] [answered]
hi!

### title(s):

* password flags for `*.zip` not took by `7z` from `Expand-7zipArchive`
* password flags not took by `7z` from `Expand-7zipArchive`
* debug *install scripts in manifests

### details
i am trying to create manifest for personal use. 
the downloaded file is a .zip file.
and for some reason, it seems that the switches from `Expand-7zipArchive` are not reaching `7z.exe` and hence, it is failing on password.

Following is what i put in there in the manifest:

```json
	"pre_install": "Expand-7zipArchive \"$dir\\$fname\" -Overwrite All -Switches '-p0000 -bb3' -Removal",
```

The [`$fname`] is inspired from [`installer.script` at App-Manifests] scoop wiki 
And `nirlauncher` seems to be [only manifest using the passwords switch][p-switch search] i.e. using `7z -p`.

[`installer.script` at App-Manifests]: https://github.com/ScoopInstaller/Scoop/wiki/App-Manifests#:~:text=%24fname
[`$fname`]: https://github.com/ScoopInstaller/Scoop/blob/f93028001fbe5c78cc41f59e3814d2ac8e595724/lib/install.ps1#L52
[p-switch search]: https://github.com/search?q=user%3AScoopInstaller+path%3Abucket%2F*.json+Expand-7zipArchive+AND+-Sw+AND+-p&type=code "user:ScoopInstaller path:bucket/*.json Expand-7zipArchive AND -Sw AND -p"

### logs
Anyhow, on running `scoop install <app>.json` it's always resulting in same `7zip.log` in the `<app>\<original_ver>` dir:

<details><summary>Details</summary>
<p>

```console
$ scoop install <app>.json 
WARN  Purging previous failed installation of <app>.
ERROR '<app>' isn't installed correctly.
Removing older version (<ver>).
'<app>' was uninstalled.
...
Installing '<app>' (<ver>) [64bit]
Loading <ar-name.zip> from cache.
WARN  Warning: No hash in manifest. SHA256 for '<app>#<ver>#<dashed-url><ar-name.zip>' is:
    2541222610927098dc0ab96603b288707e61cd4014862fb20e8f4bde611563ea
Extracting <ar-name.zip> ... ERROR Exit code was 2!
Failed to extract files from <scoop-apps>\<app>\<ver>\<ar-name.zip>.
Log file:
  <scoop-apps>\<app>\<ver>\7zip.log

Please try again or create a new issue by using the following link and paste your console output:
https:////
```

```
7-Zip 23.01 (x64) : Copyright (c) 1999-2023 Igor Pavlov : 2023-06-20

Scanning the drive for archives:
1 file, ...... bytes (... KiB)

Extracting archive: <scoop-apps-dir>\<app>\<original_ver>\<ar-name.zip>
--
Path = <same as "extracting archive">
Type = zip
Physical Size = ......


Enter password (will not be echoed):

Archives with Errors: 1


ERROR: Unspecified error

```

</p>
</details> 

whereas on manually running explicit `7z x ... -p0000` command it succeeded with this output on CLI:

<details><summary>Details</summary>
<p>

```console
$ pwd
<scoop-apps-dir>\<app>\<original_ver> 

$ 7z x ./<ar-name.zip> -p0000 -bb0

7-Zip 23.01 (x64) : Copyright (c) 1999-2023 Igor Pavlov : 2023-06-20

Scanning the drive for archives:
1 file, ...... bytes (... KiB)

Extracting archive: .\<ar-name.zip>
--
Path = <same as "extracting archive">
Type = zip
Physical Size = ......

Everything is Ok

Files: 2
Size:       ......
Compressed: ......
```

</p>
</details> 

I also tried to isolate the `scoop/lib/decompress.ps1` `Expand-7zipArchive` function call as instructed in answers here: [SO: q/1405750], but it failed as the context or say global config vars were not available to it:

[SO: q/1405750]: https://stackoverflow.com/questions/1405750/a#1406126 "call a powershell script's function directly from cli"

<details><summary>Details</summary>
<p>

```console
> Set-Variable spf -value "<scoop-apps>\scoop\current\lib"
> powershell -command "& {. $spf/decompress.ps1; Expand-7zipArchive}"

cmdlet Expand-7zipArchive at command pipeline position 1
Supply values for the following parameters:
Path: ./<ar-name.zip>
get_config : The term 'get_config' is not recognized ...

strip_ext : The term 'strip_ext' is not recognized ...

Invoke-ExternalCommand : The term 'Invoke-ExternalCommand' is not recognized ...

friendly_path : The term 'friendly_path' is not recognized ...

new_issue_msg : The term 'new_issue_msg' is not recognized ...

abort : The term 'abort' is not recognized ...

```

</p>
</details> 

[commit-7z]: https://github.com/niheaven/scoop/commit/ab1d057df9cad1a2df7da2c48912b60e38df24f0 ""
[3700]: https://github.com/ScoopInstaller/Scoop/issues/3700 "Scoop installs 7-Zip even when 7ZIPEXTRACT_USE_EXTERNAL is true"

### versions

```console
$ scoop --version | head -n2
Current Scoop version:
v0.3.1 - Released at 2022-11-15

```

**Comment by goyalyashpal** [comment]
@ rashil2000 @ niheaven 
hi! can you help 😅

**Reply by niheaven** [reply]
dot load the necessary files and test the script. It should pass the password IMO.

**Reply by goyalyashpal** [reply]
by "dot load the necessary file" i am assuming you are saying to load files from current directory (`.`). 
but how to do that? 

and what more, this time it is getting stuck at `Extracting <file> ...` and not proceeding at all

```console
$ scoop --version
Current Scoop version:
d285bb08 (HEAD -> master, tag: v0.4.1, origin/master, origin/HEAD) chore(release): Bump to version 0.4.1 (#5924)

```

here's the manifest i made:`

<details><summary>Details</summary>
<p>

```json
{
    "$schema": "https://raw.githubusercontent.com/ScoopInstaller/Scoop/master/schema.json",
    "description": "Display Real-Time Keystrokes and Mouse-Clicks on Screen",
    "homepage": "http://aalapshah.in/qipress",
    "license": "Shareware",
    "shortcuts": [
        [
            "QiPress.exe",
            "QiPress"
        ]
    ],
    "pre_install": "Expand-7zipArchive \"$dir\\qipressportable_password0000.zip\" -Overwrite All -Switches '-p0000 -bb3' -Removal",
    "persist": [
        "backgrounds",
        "resources",
        "user"
    ],
    "notes": "CheckVer scraps website, is unreliable & might show outdated number.",
    "checkver": {
        "url": "http://aalapshah.in/qipress",
        "regex": "<div class=\"field--label\">Version</div>\\s*?<div class=\"field--item\">([\\d.]+)</div>"
    },
    "version": "1.077.2",
    "architecture": {
        "64bit": {
            "url": "http://aalapshah.in/download/qipressportable_password0000.zip",
            "hash": "2541222610927098dc0ab96603b288707e61cd4014862fb20e8f4bde611563ea"
        }
    }
}

```

</p>
</details>

**Reply by niheaven** [reply]
`. xxxx\lib\decompress.ps1`

Ref: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_operators?view=powershell-7.4#dot-sourcing-operator-

**Reply by goyalyashpal** [reply]
nope, showed same errors:

```console
> # . $spf\lib\decompress.ps1; Expand-7zipArchive .\qipressportable_password0000.zip
> . $spf\lib\decompress.ps1; Expand-7zipArchive .\qipressportable_password0000.zip -Switches '-p0000'
get_config : The term 'get_config' is not recognized as the name of a cmdlet,
function, script file, or operable program. Check the spelling of the name, or
if a path was included, verify that the path is correct and try again.
At D:\UserFiles\scoop\apps\scoop\current\lib\decompress.ps1:21 char:10
+     if ((get_config USE_EXTERNAL_7ZIP)) {
+          ~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (get_config:String) [], CommandN
   otFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

strip_ext : The term 'strip_ext' is not recognized as the ....
```

**Reply by niheaven** [reply]
You should . load core.ps1 too

**Reply by goyalyashpal** [reply]
<details open><summary>tries</summary>
<p>

```powershell
cd $env:DUF\scoop\apps\qipress\1.077.2\
$spf = "$env:DUF\scoop\apps\scoop\current"
```

```powershell
. $spf\lib\decompress.ps1 $spf\lib\core.ps1; Expand-7zipArchive .\qipressportable_password0000.zip -Switches '-p0000'
```

Same error as above

```
. $spf\lib\core.ps1 $spf\lib\decompress.ps1; Expand-7zipArchive .\qipressportable_password0000.zip -Switches '-p0000'
```

Powershell crashes

</p>
</details>

**Reply by niheaven** [reply]
```pwsh
. $spf\lib\decompress.ps1
. $spf\lib\core.ps1
```

**Comment by goyalyashpal** [comment]
> ```powershell
> . $spf\lib\decompress.ps1
> . $spf\lib\core.ps1
> ```
>
> _\- @ niheaven at https://github.com/ScoopInstaller/Scoop/discussions/5866#discussioncomment-9394248_

powershell crashes:

(copy following blocks, paste in powershell)

```powershell
cd $env:DUF\scoop\apps\qipress\1.077.2\
$spf = "$env:DUF\scoop\apps\scoop\current"

. $spf\lib\decompress.ps1
. $spf\lib\core.ps1
```

```powershell
Expand-7zipArchive .\qipressportable_password0000.zip -Switches '-p0000'
```

i tried `tee-object` cmdlet, or even `tee.exe --ignore-interrupts` from msys2, but they don't survive a crash. 
so, i resorted to `wezterm record` and `wezterm replay` which revealed the error to be:

```
ERROR Exit code was 2!
new_issue_msg : The term 'new_issue_msg' is not recognized as the name of a
cmdlet, function, script file, or operable program. Check the spelling of the
name, or if a path was included, verify that the path is correct and try again.
At D:\UserFiles\scoop\apps\scoop\current\lib\decompress.ps1:47 char:95
+ ... ath.`nLog file:`n  $(friendly_path $LogPath)`n$(new_issue_msg $app $b ...
+                                                     ~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (new_issue_msg:String) [], Comma
   ndNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

Failed to extract files from .\qipressportable_password0000.zip.
Log file:
  .\7zip.log
```


<img src="https://github.com/ScoopInstaller/Scoop/assets/19423063/3ecbe741-21eb-4197-87d4-bf61971b0a7f" width=650px />

**Reply by niheaven** [reply]
Then what's in your 7zip.log? It's strange here.

**Reply by goyalyashpal** [reply]
oh i never read till now. it's funny & strange too "can't find file" 
also the curious thing, that the log file went to the root of the drive `D:\` rather than ending up in that `qipress/..../` dir

```log
7-Zip 23.01 (x64) : Copyright (c) 1999-2023 Igor Pavlov : 2023-06-20

Scanning the drive for archives:

ERROR: The system cannot find the file specified.
.\qipressportable_password0000.zip



System ERROR:
The system cannot find the file specified.
```

**Reply by niheaven** [reply]
Hmm, it's the `./` prefix that cause the problem. `Expand-7zipArchive` is designed to be used with full path, so could you try with full path again?

**Reply by goyalyashpal** [reply]
> could you try with full path again?

yes it works. surprising to me that even `.\` prefix also causes problem

```console
> lsd
qipressportable_password0000.zip

> Expand-7zipArchive D:\UserFiles\scoop\apps\qipress\1.077.2\qipressportable_password0000.zip -Switches '-p0000'

> lsd --icon=never
QiPress.exe  qipressportable_password0000.zip  README.md
```

**Comment by goyalyashpal** [comment]
> it is getting stuck at `Extracting <file> ...` and not proceeding at all
> \- @ me at https://github.com/ScoopInstaller/Scoop/discussions/5866#discussioncomment-9392161

this is continuing happening


<details><summary>manifest used</summary>
<p>

```json
{
    "$schema": "https://raw.githubusercontent.com/ScoopInstaller/Scoop/master/schema.json",
    "description": "Display Real-Time Keystrokes and Mouse-Clicks on Screen",
    "homepage": "http://aalapshah.in/qipress",
    "license": "Shareware",
    "shortcuts": [
        [
            "QiPress.exe",
            "QiPress"
        ]
    ],
    "pre_install": "Expand-7zipArchive \"$dir\\$fname\" -Switches \"-p0000\"",
    "notes": "CheckVer scraps website, is unreliable & might show outdated number.",
    "checkver": {
        "url": "http://aalapshah.in/qipress",
        "regex": "<div class=\"field--label\">Version</div>\\s*?<div class=\"field--item\">([\\d.]+)</div>"
    },
    "autoupdate": {
        "url": "http://aalapshah.in/download/qipressportable_password0000.zip"
    },
    "version": "1.077.3",
    "url": "http://aalapshah.in/download/qipressportable_password0000.zip",
    "hash": "2541222610927098dc0ab96603b288707e61cd4014862fb20e8f4bde611563ea"
}
````

</p>
</details> 


https://github.com/ScoopInstaller/Scoop/assets/19423063/bb63c9b7-e463-4e9a-9d04-3d58a8bc70fb



lack of comments support in json-manifest files make rapid trials while writing them extremely cumbersome & tedious 😭
* https://github.com/ScoopInstaller/Scoop/discussions/5913

**Reply by niheaven** [reply, marked as answer]
```json
{
    "description": "Display Real-Time Keystrokes and Mouse-Clicks on Screen",
    "homepage": "http://aalapshah.in/qipress",
    "license": "Shareware",
    "shortcuts": [
        [
            "QiPress.exe",
            "QiPress"
        ]
    ],
    "pre_install": "Expand-7zipArchive \"$dir\\$fname\" -Switches -p0000",
    "notes": "CheckVer scraps website, is unreliable & might show outdated number.",
    "checkver": {
        "url": "http://aalapshah.in/qipress",
        "regex": "<div class=\"field--label\">Version</div>\\s*?<div class=\"field--item\">([\\d.]+)</div>"
    },
    "autoupdate": {
        "url": "http://aalapshah.in/download/qipressportable_password0000.zip#/dl.zip_"
    },
    "version": "1.077.3",
    "url": "http://aalapshah.in/download/qipressportable_password0000.zip#/dl.zip_",
    "hash": "2541222610927098dc0ab96603b288707e61cd4014862fb20e8f4bde611563ea"
}
```

This is due to scoop's auto extraction of zip file, so you should *skip* the feature by renaming the downloaded file to `*.zip_`.

**Reply by goyalyashpal** [reply]
ahwwwww wowww.... 
thanks a lot. i will try it soon.

update:
it works. finalllly. i've been trying for quite long to get this working. 

and so glad that qipress stores its create on first-run config data in directories as otherwise, it it were in root-files, it would have been yet another hurdle lol

> Some programs create their config file on the first startup. Like smplayer (See: https://github.com/ScoopInstaller/Extras/pull/429)
> Currently, a file that doesn't exist in the apps directory after installation can't be made persistent.
It will create a directory named like the nonexistent file and link it to the apps directory.
>
> _\- @ r15ch13 at https://github.com/ScoopInstaller/Scoop/issues/1372#issuecomment-304462955_
